app/src/main/java/nodomain/freeyourgadget/gadgetbridge/contentprovider/HRContentProvider.java

```diff
@@ -140,85 +138,130 @@ public Cursor query(@NonNull Uri uri, String[] projection, String selection, Str
-
-                mc = new MatrixCursor(HRContentProviderContract.deviceColumnNames);
-                for (GBDevice dev : l) {
-                    mc.addRow(new Object[]{dev.getName(), dev.getModel(), dev.getAddress()});
-                }
-                return mc;
+                LOG.info("Get DEVICES LIST");
+                return devices_list();
+
             case ACTIVITY_START:
+                LOG.info("Get ACTIVTY START");
+                return activity_start(projection, selectionArgs);
 
-                this.state = provider_state.CONNECTING;
-                LOG.info(HRContentProvider.class.getName(), "Get ACTIVTY START");
-                GBDevice targetDevice = getDevice((selectionArgs != null) ? selectionArgs[0] : "");
-                if (targetDevice != null && targetDevice.isConnected()) {
-                    this.state = provider_state.ACTIVE;
-                    GBApplication.deviceService().onEnableRealtimeSteps(true);
-                    GBApplication.deviceService().onEnableRealtimeHeartRateMeasurement(true);
-                    mc = new MatrixCursor(HRContentProviderContract.activityColumnNames);
-                    mc.addRow(new String[]{"OK", "Connected"});
-                } else {
-                    GBApplication.deviceService().connect(targetDevice);
-                    mc = new MatrixCursor(HRContentProviderContract.activityColumnNames);
-                    mc.addRow(new String[]{"OK", "Connecting"});
-                }
-
-                return mc;
             case ACTIVITY_STOP:
-                this.state = provider_state.INACTIVE;
-
-                LOG.info(HRContentProvider.class.getName(), "Get ACTIVITY STOP");
-                GBApplication.deviceService().onEnableRealtimeSteps(false);
-                GBApplication.deviceService().onEnableRealtimeHeartRateMeasurement(false);
-                mc = new MatrixCursor(HRContentProviderContract.activityColumnNames);
-                mc.addRow(new String[]{"OK", "No error"});
-                return mc;
+                LOG.info("Get ACTIVITY STOP");
+                return activity_stop(projection, selectionArgs);
+
             case REALTIME:
-                //String sample_string = (buffered_sample == null) ? "" : buffered_sample.toString();
-                //LOG.error(HRContentProvider.class.getName(), String.format("Get REALTIME buffered sample %s", sample_string));
-                mc = new MatrixCursor(HRContentProviderContract.realtimeColumnNames);
-                if (buffered_sample != null)
-                    mc.addRow(new Object[]{"OK", buffered_sample.getHeartRate(), buffered_sample.getSteps(), mGBDevice != null ? mGBDevice.getBatteryLevel() : 99});
-                return mc;
+                LOG.info("REALTIME");
+                return realtime(projection, selectionArgs);
         }
         return null;
     }
 
+    @Nullable
+    private Cursor devices_list() {
```

Comment:
Suggestion for a better method name (and please try to use camel case): getDevicesCursor()? Or just getDevices()?